aptos-move/framework/supra-framework/sources/automation_registry_state.move

```diff
@@ -0,0 +1,623 @@
+            expiry_time,
+            max_gas_amount,
+            gas_price_cap,
+            state: PENDING,
+            registration_epoch,
+            registration_time,
+            tx_hash,
+        };
+
+        enumerable_map::add_value(&mut registry_data.tasks, task_index, automation_task_metadata);
+        registry_data.current_index = registry_data.current_index + 1;
+        event::emit(automation_task_metadata);
+    }
+
+    /// Cancel Automation task with specified id.
+    /// If the task was active its state is updated to be CANCELLED. Otherwise task is removed form the list.
+    /// Committed gas-limit is updated accordingly.
+    /// Only existing task can be cancled and only by task onwer.
+    public (friend) fun cancel_task(owner: &signer, id: u64): AutomationTaskMetaData acquires AutomationRegistryState {
+        let state = borrow_global_mut<AutomationRegistryState>(@supra_framework);
+        assert!(enumerable_map::contains(&state.tasks, id), EAUTOMATION_TASK_NOT_FOUND);
+
+        let automation_task_metadata = enumerable_map::get_value(&state.tasks, id);
+        assert!(automation_task_metadata.owner == signer::address_of(owner), EUNAUTHORIZED_TASK_OWNER);
+        assert!(automation_task_metadata.state != CANCELLED, EINVALID_CANCELLATION);
+        if (automation_task_metadata.state == PENDING) {
+            enumerable_map::remove_value(&mut state.tasks, id);
+        } else if (automation_task_metadata.state == ACTIVE) {
+            automation_task_metadata.state = CANCELLED;
+            enumerable_map::update_value(&mut state.tasks, id, automation_task_metadata);
+        };
+
+        // Adjust the gas committed for the next epoch by subtracting the gas amount of the cancelled task
+        state.gas_committed_for_next_epoch = state.gas_committed_for_next_epoch - automation_task_metadata.max_gas_amount;
+
+        event::emit(CanclledAutomationTask { id: automation_task_metadata.id });
+        automation_task_metadata
+    }
+
+    /// Update Automation gas limit
+    public (friend) fun update_automation_gas_limit(
+        supra_framework: &signer,
+        automation_gas_limit: u64
+    ) acquires AutomationRegistryState {
+        system_addresses::assert_supra_framework(supra_framework);
+
+        let state = borrow_global_mut<AutomationRegistryState>(@supra_framework);
+        assert!(state.gas_committed_for_next_epoch < automation_gas_limit, EUNACCEPTABLE_AUTOMATION_GAS_LIMIT);
+
+        state.automation_gas_limit = automation_gas_limit;
+
+        event::emit(UpdateAutomationGasLimit { automation_gas_limit });
+    }
+
+    /// List all the automation task ids
+    public(friend) fun get_active_task_ids(): vector<u64> acquires AutomationRegistryState {
+        let state = borrow_global<AutomationRegistryState>(@supra_framework);
+
+        let active_task_ids = vector[];
+        let ids = enumerable_map::get_map_list(&state.tasks);
+
+        vector::for_each(ids, |id| {
```

Comment:
Since we are only reading the state, perhaps `vector::for_each_ref` can be used to avoid creation/copy of elements.

Reply:
@nizam-supraoracles